### DISCUSSION TITLE
Load Balancing / High Availability

### DISCUSSION THREAD

**Discussion by dnachev** [immich-app/immich#1762] [Q&A]
I was wondering are there any concerns for running multiple containers if the database, the redis cluster and the upload folder is properly setup.

I am running immich services with Docker Swarm and I'd like to be able to spread the load of the server and the respective tasks across more than one machine. It seems the architecture is prepared for this, but it is unclear whether it is actually supported/validated.

**Comment by bo0tzz** [comment]
It should be fine to scale the services up. The microservices container does the most work, so that one will be the most impactful.

**Comment by AlmightyFrog** [comment]
Trying to collect information here about how scaling could be done.
I want to understand the constraints to see what can be done especially when it comes to temporary load balancing (powering on other systems on demand).
In the following there are some assumptions which might be totally wrong from my side.

---

**immich-server**
- Does depend on UPLOAD_LOCATION or external libraries mounted
- Scaling does not make much sense, single instance sufficient
- library/encoded-video and library/thumbs best put on SSD as they are usually always used while browsing and looking at picures
- UPLOAD_LOCATION or external libraries can be put on slow or network based storage as only used when scan changes or process new files

**immich-microservices**
- Does depend on UPLOAD_LOCATION or external libraries mounted
- Depends on immich-postgres, so this would be needed to make accessible over network
- Scaling would be possible e.g. using nginx if filesystem is shared.

**immich-machine-learning**
- Does not depend on UPLOAD_LOCATION or external libraries mounted
- Can be easily added/removed any time e.g. using nginx as load balancer (see #4616)

---

Motivation:
- As my homeserver is build to be power efficient it does not have a dedicated GPU build in (sadly video endoding with quicksync on iGPU seems not to work)
- I am running out of fast disk space. Seeing now what a full rebuild of my assets takes time (about 168000 photos and 3300 videos currently) i am considering restructuring everything to future proof it more. But I do of cause not want to spin up the big hard drive too often cause immich wants to check something.

See also: #4616, #1862

**Reply by AlmightyFrog** [reply]
Current status:

*immich-microservices*: Not sure how to go further here. I did get immich-microservices running on other host with access to original filesystem and database, but can't find any solution to make my main immich-server to connect to it. Seems like host "immihc -microservices" is hardcoded. Any suggestion @bo0tzz or @dnachev how to get further without using docker swarm?

*immich-machine-learning*: There was no issue to put this on another host with a dedicated GPU.

**Reply by bo0tzz** [reply]
> make my main immich-server to connect to it

They don't connect directly. As long as microservices has access to redis, postgres, and the filesystem, it'll take jobs from redis and process them.

**Reply by AlmightyFrog** [reply]
Note to anyone who might read this and is interested in similar topic: See also #7925 for good further information.
